tdiag/src/commands/graph.rs

```diff
@@ -0,0 +1,232 @@
+
+use crate::DiagError;
+
+use timely::dataflow::operators::{Filter, capture::{Capture, extract::Extract}};
+use timely::dataflow::operators::map::Map;
+
+use differential_dataflow::collection::AsCollection;
+use differential_dataflow::operators::{Join, reduce::Threshold, Consolidate};
+
+use timely::logging::TimelyEvent::{Operates, Channels};
+
+use tdiag_connect::receive::ReplayWithShutdown;
+
+static GRAPH_HTML: &str = include_str!("graph/dataflow-graph.html");
+
+/// Creates an html file that renders a timely dataflow graph.
+///
+/// 1. Listens to incoming connection from a timely-dataflow program with
+/// logging enabled;
+/// 2. runs a differential-dataflow program to reconstruct the source
+/// computation's dataflow graph;
+/// 3. writes an html file that renders the dataflow graph.
+///
+/// This module includes `graph/dataflow-graph.html` as a static resource.
+pub fn listen_and_render(
+    timely_configuration: timely::Configuration,
+    sockets: Vec<Option<std::net::TcpStream>>,
+    output_path: &std::path::Path) -> Result<(), crate::DiagError> {
+
+    let sockets = Arc::new(Mutex::new(sockets));
+
+    let (operators_send, operators_recv) = ::std::sync::mpsc::channel();
+    let operators_send = Arc::new(Mutex::new(operators_send));
+
+    let (channels_send, channels_recv) = ::std::sync::mpsc::channel();
+    let channels_send = Arc::new(Mutex::new(channels_send));
+
+    let is_running = Arc::new(std::sync::atomic::AtomicBool::new(true));
+    let is_running_w = is_running.clone();
+
+    let worker_handles = timely::execute(timely_configuration, move |worker| {
+        let operators_send: std::sync::mpsc::Sender<_> = operators_send.lock().expect("cannot lock operators_send").clone();
+        let channels_send: std::sync::mpsc::Sender<_> = channels_send.lock().expect("cannot lock channels_send").clone();
+
+        let sockets = sockets.clone();
+
+        // create replayer from disjoint partition of source worker identifiers.
+        let replayer = tdiag_connect::receive::make_readers::<
+            std::time::Duration, (std::time::Duration, timely::logging::WorkerIdentifier, timely::logging::TimelyEvent)>(
+            tdiag_connect::receive::ReplaySource::Tcp(sockets), worker.index(), worker.peers())
+            .expect("failed to open tcp readers");
+
+        worker.dataflow(|scope| {
+            let stream = replayer.replay_with_shutdown_into(scope, is_running_w.clone())
+                .filter(|(_, worker, _)| *worker == 0);
+
+            let operates = stream
+                .flat_map(|(t, _, x)| if let Operates(event) = x { Some((event, t, 1 as isize)) } else { None })
+                .as_collection();
+
+            let channels = stream
+                .flat_map(|(t, _, x)| if let Channels(event) = x { Some((event, t, 1 as isize)) } else { None })
+                .as_collection();
+
+            // == Fix addresses so we can connect operators outside and inside subgraphs ==
```

Comment:
I think it would be worthwhile to add a few more sentences of explanation here, as I think by now four different people have reimplemented this logic one way or the other. E.g. operates events are actually a union of actual operators and scopes, and that because operator addresses are assigned hierarchically, `addr.pop()` will give the scope that an operator is contained in. We want to only render operators as nodes in the graph, but need the scoping information to box them correctly and to highlight edges crossing scope boundaries.

In that same vein it might be clearer to rename `operates_anti` => `scopes`?

Reply:
Good point, forgot about that! I'm going to merge this, then work on adding some explanation of the logic.